app/src/main/java/org/p2p/wallet/push_notifications/ineractor/PushNotificationsInteractor.kt

```diff
@@ -35,7 +41,20 @@ class PushNotificationsInteractor(
             clientId = tokenKeyProvider.publicKey
         )
 
-        return deviceTokenRepository.sendDeviceToken(deviceToken)
+        try {
+            deviceTokenRepository.sendDeviceToken(deviceToken)
+        } catch (e: ServerException) {
+            retryTokenSend(retries)
+        }
+    }
+
+    private fun retryTokenSend(retries: Int) {
+        if (retries < 1) return
+
+        appScope.launch {
+            delay(TOKEN_SEND_RETRY_DELAY)
+            updateDeviceToken(retries - 1)
+        }
     }
```

Comment:
```suggestion
        appScope.launch {
            sendTokenToServer(deviceToken, retries)
        }
    }

    private suspend fun sendTokenToServer(deviceToken: DeviceToken, retries: Int) {
        val noRetryNeeded = retries == 0
        if (noRetryNeeded) return

        try {
            deviceTokenRepository.sendDeviceToken(deviceToken)
        } catch (e: ServerException) {
            delay(TOKEN_SEND_RETRY_DELAY)
            sendTokenToServer(deviceToken, retries - 1)
        }
    }
```

Reply:
your first send try was not in appScope, so I moved it
i extracted sending method so it's recursion in only one place, plus no shared preferences logic included any longer

Reply:
Thank, that'a a good point!